jb/detail/array_traits.hpp

```diff
@@ -0,0 +1,73 @@
+#ifndef jb_detail_array_traits_hpp
+#define jb_detail_array_traits_hpp
+
+#include <boost/multi_array.hpp>
+#include <vector>
+
+namespace jb {
+namespace detail {
+/**
+ * Define traits for array-like containers.
+ *
+ * The generic version implements the class assuming std::vector<> or
+ * any similar container.  We provide specializations for other types
+ * (such as boost::multi_array<>) to be able to operate on them
+ * generically.
+ *
+ * @tparam container_type the type of container, typically std::vector<>
+ */
+template <typename container_type>
+struct array_traits {
+  /// Define the type of the elements in the container
+  using element_type = typename container_type::value_type;
+};
+
+/// Count the number of elements for a vector-like container
+template <typename container_type>
+inline std::size_t element_count(container_type const& a) {
+  return a.size();
+}
+
+/// Count the elements in the last dimension
```

Comment:
Same thing, needs longer documentation, maybe model after the doc for array_traits<>.

I know this is my code, but now it is going into the master branch.